File: backend/app/models/note.py
import uuid
from datetime import datetime, timezone

from sqlalchemy import String, Text, DateTime, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Note(Base):
    __tablename__ = "notes"

    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), primary_key=True, default=uuid.uuid4
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    content_markdown: Mapped[str] = mapped_column(Text, nullable=False, default="")
    source_image_url: Mapped[str | None] = mapped_column(String(1024), nullable=True)
    ai_provider_used: Mapped[str | None] = mapped_column(String(50), nullable=True)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    user: Mapped["User"] = relationship("User", back_populates="notes")  # noqa: F821
    mind_map: Mapped["MindMap | None"] = relationship(  # noqa: F821
        "MindMap", back_populates="note", uselist=False, cascade="all, delete-orphan", lazy="selectin"
    )

    def __repr__(self) -> str:
        return f"<Note {self.title[:40]}>"

File: frontend/lib/models/note.dart
class Note {
  final String id;
  final String title;
  final String contentMarkdown;
  final String? sourceImageUrl;
  final String? aiProviderUsed;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Note({
    required this.id,
    required this.title,
    required this.contentMarkdown,
    this.sourceImageUrl,
    this.aiProviderUsed,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Note.fromJson(Map<String, dynamic> json) {
    return Note(
      id: json['id'] as String,
      title: json['title'] as String? ?? 'Untitled',
      contentMarkdown: json['content_markdown'] as String? ??
          json['contentMarkdown'] as String? ??
          '',
      sourceImageUrl: json['source_image_url'] as String? ??
          json['sourceImageUrl'] as String?,
      aiProviderUsed: json['ai_provider_used'] as String? ??
          json['aiProviderUsed'] as String?,
      createdAt: DateTime.parse(
          json['created_at'] as String? ?? json['createdAt'] as String? ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(
          json['updated_at'] as String? ?? json['updatedAt'] as String? ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'content_markdown': contentMarkdown,
      'source_image_url': sourceImageUrl,
      'ai_provider_used': aiProviderUsed,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }

  Note copyWith({
    String? id,
    String? title,
    String? contentMarkdown,
    String? sourceImageUrl,
    String? aiProviderUsed,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Note(
      id: id ?? this.id,
      title: title ?? this.title,
      contentMarkdown: contentMarkdown ?? this.contentMarkdown,
      sourceImageUrl: sourceImageUrl ?? this.sourceImageUrl,
      aiProviderUsed: aiProviderUsed ?? this.aiProviderUsed,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }
}
